### ISSUE TITLE
OpenIn Extension stopped working - current Intune compatibility with ADAL

### ISSUE THREAD

**Issue by kariohlsen** [msintuneappsdk/ms-intune-app-sdk-ios#332]
**Describe the bug:**
The Open In Action extension that we previously had working has stopped working with recent upgrades to Intune.  We upgraded from  12.9.4 to 17.4.0.  I have combed through the Intune Docs and see that MSAL is now "required".  Previously we were using ADAL v 4.0.6 and I have tried updating to v 6.0.4 but still get same issue. Is current Intune compatible with ADAL or is there another issue? Must we upgrade to MSAL to work with the new Intune version?

I have verified the app groups, keychain groups, action extension subquery and all look fine which is not surprising as all of these used to work in earlier version.

**To Reproduce**
1. Go to Word and open a document from a Shared location
2. Tap 3-dot menu and select Send a Copy
3. Select Send with Another App
4. Select Open In (myApp)

**Expected behavior:**
My App should open with the selected document.

**Observed behavior**
The Share sheet is dismissed and logs show no errors but My App is not launched

**Screenshots and logs:**

Logs from working version (Intune 12.9.4/ADAL 4.0.6)
default	17:45:28.736882-0700	Word	completed request - items: (
    "<NSExtensionItem: 0x282ced6b0> - userInfo: {\n    NSExtensionItemAttachmentsKey =     (\n        \"<NSItemProvider: 0x2805ec4d0> {types = (\\n    \\\"com.microsoft.intune.mam.itemprovider.openin.uti\\\",\\n    \\\"com.microsoft.intune.mam.itemprovider.protectionprovider.CMARPassthroughItemProtectionProvider\\\"\\n)}\"\n    );\n}"
)
default	17:45:28.736927-0700	Word	Called activityDidFinish:YES items:(
    "<NSExtensionItem: 0x282ced6b0> - userInfo: {\n    NSExtensionItemAttachmentsKey =     (\n        \"<NSItemProvider: 0x2805ec4d0> {types = (\\n    \\\"com.microsoft.intune.mam.itemprovider.openin.uti\\\",\\n    \\\"com.microsoft.intune.mam.itemprovider.protectionprovider.CMARPassthroughItemProtectionProvider\\\"\\n)}\"\n    );\n}"
) error:(null)
default	17:45:28.736970-0700	Word	Completing with state:Succeeded <UIApplicationExtensionActivity: 0x281533cf0> 
default	17:45:28.737118-0700	Word	Did perform activity with result:<SHSheetActivityPerformerResult activity:<UIApplicationExtensionActivity: 0x281533cf0> activityType:com.xxxxxx.OpenInExt activityTitle:UIApplicationExtensionActivity completedState:Succeeded returnedItems:(
    "<NSExtensionItem: 0x282ced6b0> - userInfo: {\n    NSExtensionItemAttachmentsKey =     (\n        \"<NSItemProvider: 0x2805ec4d0> {types = (\\n    \\\"com.microsoft.intune.mam.itemprovider.openin.uti\\\",\\n    \\\"com.microsoft.intune.mam.itemprovider.protectionprovider.CMARPassthroughItemProtectionProvider\\\"\\n)}\"\n    );\n}"
) error:(null)>

Logs from non-working version (Intune 17.4.0/ADAL 4.0.6 or 6.0.4)
default	17:39:51.121845-0700	Word	completed request - items: (
)
default	17:39:51.121913-0700	Word	Called activityDidFinish:YES items:(
) error:(null)
default	17:39:51.121978-0700	Word	Completing with state:Succeeded <UIApplicationExtensionActivity: 0x281a179b0> activityType:com.xxxxxx.OpenInExt activityTitle:UIApplicationExtensionActivity (com.xxxxxx.OpenInExt)
default	17:39:51.122293-0700	Word	Did perform activity with result:<SHSheetActivityPerformerResult activity:<UIApplicationExtensionActivity: 0x281a179b0> activityType:com.xxxxxxx.OpenInExt activityTitle:UIApplicationExtensionActivity completedState:Succeeded returnedItems:(
) error:(null)>

**Smartphone (please complete the following information):**
 - Device: iPhone XR and iPhone 13 Pro Max
 - OS: iOS 16.0

**Intune App SDK for iOS (please complete the following information):**
- What version of the Intune SDK are you using? Are you using the latest version?
- Intune 17.4.0
- What platform is your app based in (native, Xamarin based, Cordova, etc)?
- native iOS
- For errors during build, does the app build without Intune SDK integration?
- N/A
- For errors post build, does the app launch without being Intune SDK integrated?
- N/A

- Who is the customer?
- Do you see a trend with it only being reproduced on a specific device?

**Additional context:**

**Comment by vigavadi** [comment]
@kariohlsen It's highly recommended to move to MSAL. ADAL support is going to be completely dropped soon. 
Can you send the full set of logs both working and non working scenario? You might need to implement the IntuneMAMDiagnosticConsole API if you haven't already. Here's the link: [How can I troubleshoot my app](https://learn.microsoft.com/en-us/mem/intune/developer/app-sdk-ios#how-can-i-troubleshoot-my-app)

Once the diagnostic view controller is launched, Click on Get Started -> Send logs with Microsoft -> Send me the IncidentId

**Comment by kariohlsen** [comment]
@vigavadi Thanks.  I shared the logs and Reference ID is D00565

**Comment by kariohlsen** [comment]
I added the Intune Diagnostic console to the working version, but the only option from this old Intune version is Share (not upload to MS). Can you give me email address to send file to?

Incidentally the working flow prompted me with "My App wants to open file" dialog the first time I transferred a file. I think this is a clue..

**Comment by kariohlsen** [comment]
I am uploading another set of non-working logs for Reference ID 64C135 - here I have linked the intune libraries in the extension (did not see that in Intune refs, but did see it in old code dredged up to repro working case). I am getting an error on Word side now that says "This operation failed: Error 2".

**Comment by dtatcbrain** [comment]
We've got this bug as well. We tested various versions of the Intune SDK and it seems you introduced it in version 17.1.1. This is in the release notes for that version, so maybe it's related to that?
"Fixed issue where users could not share from an app with unrestrictive outgoing data transfer policy to an Intune-enabled app that is unmanaged via a share extension."

**Comment by vigavadi** [comment]
@kariohlsen Thanks I'll take a look. you can send the working logs to vigavadi@microsoft.com. Can you also include repro videos?

**Comment by kariohlsen** [comment]
@vigavadi I sent the logs and demo videos to your email.  BTW I am seeing a crash on Word at the end of the failed flow.

**Comment by vigavadi** [comment]
@kariohlsen Thanks! I can't see any logs for action extension. Do you have separate logs for those? Also, Can you send me logs from Edge as well. That should include logs from Word. Go to Edge -> about://intunehelp -> Send logs to Microsoft

**Comment by kariohlsen** [comment]
OK, I uploaded the logs from the non-working device.  Ref ID 6DE84C.  The only place I have seen the logs for the extension are in the console.. I will send you console logs that I have via email

**Comment by kariohlsen** [comment]
@dtatcbrain thanks so much for identifying the regression point.  I integrated Intune v 17.0.0 per your comment and Open in Extension is working for me now.

**Comment by kariohlsen** [comment]
@vigavadi is there any ETA on when a fix might be provided for this?

**Comment by vigavadi** [comment]
@kariohlsen Working on getting a repro. Your app has "Allow data transfer to other apps" set to Policy Managed or Any Apps. In the logs I see both so just wanted to confirm.

**Comment by kariohlsen** [comment]
@vigavadi Yes we have policy set to Allow data transfer to other apps = Policy Managed.  Also the com.mcrosoft.intune.mam keychain group is in the extension as well. I made no changes to the policies but was able to get the extension working again when I downgraded intune to 17.0.0

**Comment by vigavadi** [comment]
I see different App Id prefix for app and the extension in the logs.

**Comment by kariohlsen** [comment]
@vigavadi there are no changes in the app code or properties between the non-working case using Intune 17.4.0 and the working case using Intune 17.0.0.  Can you check the working case logs and send me what the discrepant app Id prefixes are by email please?

**Comment by vigavadi** [comment]
Thanks Kari, We’ll look into the logs you shared for 17.0.0. The bundle prefix isn’t MAM specific. MAM just fetches these values from the app. Would you be able to share the App and extension Info.plist?
 
Is this app available on App store? Or is there a way to share the app and a test user? This would help with investigations. Just to double check does this happen every time you try to share a file from Word to your app or is that for specific documents? Does it happen with any other app like OneDrive or Outlook?

**Comment by kariohlsen** [comment]
@vigavadi The problem also exists with OneNote.  I get the message "This operation failed. Error 2."  The app is available on App Store but only with the earlier working version of Intune of course. I will send you plists and app info by email

**Comment by gastaffo** [comment]
Hi @kariohlsen. Vidhi is out for this week. I can take a look at your case while she is gone. Can you send those plists and app info to gastaffo@microsoft.com. Or you can wait until Vidhi is back if you prefer.

**Comment by kariohlsen** [comment]
@gastaffo to summarize the history here.. the current app configuration works fine for Intune version 12.9.4 and also 17.0.0.  I only see an issue when I upgrade to Intune v 17.4.0. Another user @dtatcbrain identified the regression point in an earlier post here. 

I will send you screenshots you requested by email.  I have gone through the supplied links multiple times and found no issues on my side, and I believe that the fact it works with earlier Intune versions validates this.

**Comment by dtatcbrain** [comment]
"Someone else" was actually us - turns out we had missed step 3 here, and adding keychain sharing to the extension solved the issue. 
https://learn.microsoft.com/en-us/mem/intune/developer/app-sdk-ios#copy-to-actions

This also surprised us since it worked before 17.1.1, but apparently this step had no effect until then.

**Comment by kariohlsen** [comment]
Thanks @dtatcbrain but my keychain sharing setup in extension seems fine. I re-sent working and nonworking logs to @gastaffo and @vigavadi on Friday. Please let me know what else you need to solve this. I am working on getting an account to send you for testing the TestFlight version I will provide.

**Comment by vigavadi** [comment]
@kariohlsen Thanks! I'll take a look.

**Comment by kariohlsen** [comment]
OK!  I finally have the solution.  I needed to fix the provisioning profile generated on the Apple site for the extension.  It turns out that you select the app prefix when you create the app id for the Action extension and there were several to choose from on my team.  The person who created the app id used the default it looks like (first one on the list).  I created a new app id for the extension using the proper app prefix, and I am now able to perform Open In action with Intune 17.4.0. As pointed out by @dtatcbrain this was unexpected as it worked in 17..0.0 and prior, but this is the root cause. This post helped me here https://github.com/msintuneappsdk/ms-intune-app-sdk-ios/issues/63#top. Thanks for your help @vigavadi @gastaffo
